### PR TITLE
fix(pnpr): preserve non-canonical upstream tarball filenames

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

The proxied-tarball integrity fix (pnpm/pnpm#12570, GHSA-5f9g-98vq-2jxw) started reconstructing each rewritten `dist.tarball` filename from its version key (`<name>-<version>.tgz`). That assumes upstream tarball filenames are always canonical, which breaks packages whose real tarball name differs — e.g. `esprima-fb@3001.1.0-dev-harmony-fb`, whose npm tarball is `esprima-fb-3001.0001.0000-dev-harmony-fb.tgz`. The client recorded the wrong lockfile URL, and the proxied fetch 404'd against the upstream.

This regressed two tests on `main` (`@pnpm/installing.deps-installer` → `test/lockfile.ts`):

- `save tarball URL when it is non-standard`
- `keep non-standard tarball URL when lockfileIncludeTarballUrl is false`

This PR restores faithful proxying of non-canonical tarball names while keeping the GHSA integrity protection intact.

## Squash Commit Body

```text
fix(pnpr): preserve non-canonical upstream tarball filenames

The proxied-tarball integrity fix (pnpm/pnpm#12570, GHSA-5f9g-98vq-2jxw)
started reconstructing each rewritten dist.tarball filename from its
version key (`<name>-<version>.tgz`). That assumes upstream tarball names
are always canonical, which breaks packages like esprima-fb whose real
npm tarball is `esprima-fb-3001.0001.0000-dev-harmony-fb.tgz` for version
`3001.1.0-dev-harmony-fb`: the rewritten URL no longer matched what npm
hosts, so the client recorded the wrong lockfile URL and the proxied
fetch 404'd.

Preserve the upstream basename verbatim in the rewrite again, and resolve
a tarball request's version and dist.integrity by matching the requested
filename against each version's dist.tarball basename instead of parsing
the version out of the filename. OSV screening re-runs against the
resolved version when it differs from the filename-derived one.

The GHSA protection is unchanged: served bytes are still verified against
the SRI declared by the version that owns that tarball name, so a
preserved name cannot smuggle in bytes of another provenance. Tests that
encoded the canonical-reconstruction behavior are updated to assert
basename preservation, with new coverage for non-canonical names.

Also fix a pre-existing compile break in the pnpr server test target:
UserStore::in_memory gained a MaxUsers argument (pnpm/pnpm#12581) but one
test call site was not updated.
```

## Checklist

- [x] The change is implemented in both the TypeScript CLI and the Rust `pacquet/` port, or the description notes what still needs porting. — N/A: the change is confined to the `pnpr` registry server; no pnpm/pacquet CLI surface is affected.
- [x] Added a changeset (`pnpm changeset`) if this PR changes any published package. — N/A: `pnpr` is a Rust binary, not a published npm package.
- [x] Added or updated tests.
- [x] Updated the documentation if needed. — N/A.

---
Written by an agent (Claude Code, claude-opus-4-8).

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved tarball routing and authorization by preserving the `dist.tarball` basename declared in the packument and ensuring integrity/provenance checks align with the declaring version, even when packuments swap filenames.

* **New Features**
  * Added more robust tarball URL rewriting that normalizes basenames consistently, including handling query/signed URLs and non-canonical paths.

* **Tests**
  * Expanded unit and server tests for tarball URL edge cases, basename extraction, and failure behavior when multiple versions share an ambiguous tarball basename.

### CODE REVIEW

File: pnpr/crates/pnpr/src/server.rs [hunk 1]
```diff
@@ -952,18 +966,35 @@ async fn serve_tarball(
...
+        return Ok(None);
+    };
+    let Some((version, manifest)) = versions.iter().find(|(_, manifest)| {
+        manifest
+            .get("dist")
+            .and_then(|dist| dist.get("tarball"))
+            .and_then(Value::as_str)
+            .and_then(|url| url.rsplit('/').next())
+            .is_some_and(|basename| basename == filename)
+    }) else {
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpr/crates/pnpr/src/server.rs]: Fixed in bd3982486e. `expected_tarball_dist` now uses `filter` instead of `find`: it takes the first basename match and then fails closed (502 via `tarball_integrity_error`) if a second version declares the same `dist.tarball` basename, so a request can never bind integrity/OSV to an iteration-order-dependent version. Added the integration test `ambiguous_tarball_basename_is_rejected_before_fetch` (no upstream fetch on the ambiguous path).

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pnpr/crates/pnpr/src/upstream.rs [hunk 2]
```diff
@@ -358,31 +358,44 @@ impl Upstream {
...
     };
     let Some(tarball_value) = dist.get_mut("tarball") else { return };
-    if !tarball_value.is_string() {
+    let Some(basename) = tarball_value.as_str().and_then(|url| url.rsplit('/').next()) else {
         return;
-    }
-    let filename = pkg.tarball_name_for_version(version);
-    *tarball_value = Value::String(format!("{public_url}/{}/-/{filename}", pkg.as_str()));
+    };
+    *tarball_value = Value::String(format!("{public_url}/{}/-/{basename}", pkg.as_str()));
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpr/crates/pnpr/src/upstream.rs]: Fixed in bd3982486e. Added a shared `upstream::tarball_basename` helper that strips `?query`/`#fragment` before taking the last path segment, and use it in both `rewrite_dist_tarball` and `expected_tarball_dist` so the rewritten public URL and the serve-time match always derive the basename identically (including for signed/query-bearing URLs). Covered by `rewrites_query_bearing_tarball_to_its_path_basename` and `tarball_basename_strips_query_and_fragment`.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pnpr/crates/pnpr/src/upstream.rs [hunk 3]
```diff
@@ -358,31 +358,56 @@ impl Upstream {
...
     let Some(tarball_value) = dist.get_mut("tarball") else { return };
-    if !tarball_value.is_string() {
+    let Some(basename) = tarball_value.as_str().and_then(tarball_basename) else {
         return;
-    }
-    let filename = pkg.tarball_name_for_version(version);
-    *tarball_value = Value::String(format!("{public_url}/{}/-/{filename}", pkg.as_str()));
+    };
+    *tarball_value = Value::String(format!("{public_url}/{}/-/{basename}", pkg.as_str()));
+}
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpr/crates/pnpr/src/upstream.rs:415]: Good catch — fixed in 4559f27ff3. `rewrite_dist_tarball` now always rewrites a string `dist.tarball` to a route on this server: it prefers the URL's basename, falls back to the version-derived canonical name (`tarball_name_for_version`, from the manifest's own `version`), and finally an empty filename — so a URL with no usable basename (e.g. a trailing slash) can never be passed through to the client. A basename-less URL therefore resolves to a pnpr route that 404s at `serve_tarball` rather than directing the client at an arbitrary upstream host. Covered by the new test `rewrites_basenameless_tarball_url_to_pnpr_route`.

---
Written by an agent (Claude Code, claude-opus-4-8).
